File: server/routes/api/issues.js
const db = require('../../models')
const localIssues = require('./local/issues')
const jiraIssues = require('./jira/issues')

module.exports = function (router) {
  router.get('/issues', function (req, res) {
    if (req.settings && req.settings.useJira) {
      return jiraIssues.findAllIssues(req, res)
    }
    return localIssues.findAllIssues(req, res)
  })

  router.get('/issues-by-epic', function (req, res) {
    if (req.settings && req.settings.useJira) {
      return jiraIssues.issuesByEpic(req, res)
    }
    return localIssues.issuesByEpic(req, res)
  })

  router.get('/issue/:issueId', function (req, res) {
    if (req.settings && req.settings.useJira) {
      return jiraIssues.get(req, res)
    }
    return localIssues.get(req, res)
  })

  router.get('/issues/search', function (req, res) {
    if (req.settings && req.settings.useJira) {
      return jiraIssues.search(req, res)
    }
    return localIssues.search(req, res)
  })

  router.get('/issues/standup', function (req, res) {
    if (req.settings && req.settings.useJira) {
      return jiraIssues.standup(req, res)
    }
    return localIssues.standup(req, res)
  })

  router.get('/issues/backlog', function (req, res) {
    if (req.settings && req.settings.useJira) {
      return jiraIssues.backlog(req, res)
    }
    return localIssues.backlog(req, res)
  })

  router.post('/issue', function (req, res) {
    if (req.settings && req.settings.useJira) {
      return jiraIssues.create(req, res)
    }
    return localIssues.create(req, res)
  })

  router.put('/issue/:issueId', function (req, res) {
    if (req.settings && req.settings.useJira) {
      return jiraIssues.update(req, res)
    }
    return localIssues.update(req, res)
  })

  router.put('/issue/:issueId/assign', function (req, res) {
    if (req.settings && req.settings.useJira) {
      return jiraIssues.assign(req, res)
    }
    return localIssues.assign(req, res)
  })

  router.put('/issue/:issueId/status/:statusId', function (req, res) {
    if (req.settings && req.settings.useJira) {
      return jiraIssues.updateStatus(req, res)
    }
    return localIssues.updateStatus(req, res)
  })

  router.delete('/issue/:issueId', function (req, res) {
    return db.Issue.destroy({ where: { id: req.params.issueId } })
  })

  router.get('/epics/search', function (req, res) {
    if (req.settings && req.settings.useJira) {
      return jiraIssues.searchEpics(req, res)
    }
    return localIssues.searchEpics(req, res)
  })

  router.get('/issue/:issueId/subtasks', function (req, res) {
    if (req.settings && req.settings.useJira) {
      return jiraIssues.getSubtasks(req, res)
    }
    return localIssues.getSubtasks(req, res)
  })

  router.get('/issue/:issueId/changelog', function (req, res) {
    if (req.settings && req.settings.useJira) {
      return jiraIssues.getChangeLog(req, res)
    }
    return localIssues.getChangeLog(req, res)
  })
}

File: server/api/issues.py
from flask import g, jsonify, request
from repository import repo_factory, RepositoryTypes
from .app import API_APP


@API_APP.route('/issues', methods=["GET"])
def get_issues():
    repo = repo_factory(g, RepositoryTypes.ISSUE)
    results = repo.get_in_progress()
    return jsonify([issue.to_viewmodel() for issue in results])


@API_APP.route('/issues-by-epic', methods=["GET"])
def get_issues_by_epic():
    return ('', 200)


@API_APP.route('/issues/<int:issue_id>', methods=['GET'])
def get_issue(issue_id):
    repo = repo_factory(g, RepositoryTypes.ISSUE)
    issue = repo.get_by_id(issue_id)
    return jsonify(issue.to_viewmodel())


@API_APP.route('/issues/search', methods=["GET"])
def search_issues():
    repo = repo_factory(g, RepositoryTypes.ISSUE)
    return ('', 200)


@API_APP.route('/issues/standup', methods=["GET"])
def get_standup_issues():
    repo = repo_factory(g, RepositoryTypes.ISSUE)
    return ('', 200)


@API_APP.route('/issues/backlog', methods=["GET"])
def get_backlog_issues():
    repo = repo_factory(g, RepositoryTypes.ISSUE)
    results = repo.get_backlog_issues()
    return jsonify([issue.to_viewmodel() for issue in results])


@API_APP.route('/issues', methods=['POST'])
def create_issue():
    repo = repo_factory(g, RepositoryTypes.ISSUE)
    issue = repo.create(request.json, g.current_user.id)
    return jsonify(issue.to_viewmodel())


@API_APP.route('/issues/<int:issue_id>', methods=['PUT'])
def update_issue(issue_id):
    repo = repo_factory(g, RepositoryTypes.ISSUE)
    issue = repo.update_issue(request.json, g.current_user.id)
    return jsonify(issue.to_viewmodel())


@API_APP.route('/issues/<int:issue_id>/assign', methods=['PUT'])
def assign_issue(issue_id):
    repo = repo_factory(g, RepositoryTypes.ISSUE)
    assignee = request.json if request.json else {}
    issue = repo.assign(assignee.get('id'), issue_id, g.current_user.id)
    return jsonify(issue.to_viewmodel())


@API_APP.route('/issues/<int:issue_id>/status/<int:status_id>', methods=['PUT'])
def update_issue_status(issue_id, status_id):
    repo = repo_factory(g, RepositoryTypes.ISSUE)
    issue = repo.update_status(issue_id, status_id, g.current_user.id)
    return jsonify(issue.to_viewmodel())


@API_APP.route('/issues/<int:issue_id>', methods=['DELETE'])
def delete_issue(issue_id):
    repo = repo_factory(g, RepositoryTypes.ISSUE)
    repo.delete(issue_id)
    return ('', 204)


@API_APP.route('/issues/<int:issue_id>/subtasks', methods=["GET"])
def get_issue_subtasks(issue_id):
    repo = repo_factory(g, RepositoryTypes.ISSUE)
    results = repo.get_children(issue_id)
    return jsonify([issue.to_viewmodel() for issue in results])


@API_APP.route('/issues/<int:issue_id>/changelog', methods=["GET"])
def get_issue_changelog(issue_id):
    repo = repo_factory(g, RepositoryTypes.CHANGE_LOG)
    history = repo.get_changelog_for_issue(issue_id)
    return jsonify([log.to_viewmodel() for log in history])
